colossalai/utils/model/pipelinable.py

```diff
@@ -0,0 +1,244 @@
+                for layer_spec in self._layer_spec_list:
+                    param_counts.append(layer_spec.count_params())
+                parts = partition_balanced(param_counts, pipeline_size, num_chunks)[rank]
+            else:
+                raise ValueError("A string partition policy should be one of ['uniform', 'balanced'].")
+        elif isinstance(self._policy, dict):
+            parts = self._policy[rank]
+        else:
+            raise ValueError("A partition policy should be either a string or a dictionary.")
+
+        layers_to_build = []
+        for start, end in parts:
+            layers_to_build += self._layer_spec_list[start:end]
+        func_dict_in_partition = {}
+        module_list_in_partition = []
+        if rank == 0 and "first" in self._func_dict:
+            func_dict_in_partition["first"] = self._func_dict["first"]
+        for layer in layers_to_build:
+            module = layer.build()
+            module_list_in_partition.append(module)
+            if layer in self._func_dict:
+                func_dict_in_partition[id(module)] = self._func_dict[layer]
+        module_list_in_partition = torch.nn.ModuleList(module_list_in_partition)
+        pipeline_model = PipelinableModel(module_list_in_partition, func_dict_in_partition)
+
+        return pipeline_model
+
+    def load_policy(self, policy):
+        self._policy = policy
+
+
+class PipelinableModel(torch.nn.Module):
+
+    def __init__(self, module_list, func_dict):
+        super().__init__()
+        self._module_list = module_list
+        self._func_dict = func_dict
+
+    def forward(self, input_tensor):
+        if "first" in self._func_dict:
+            funcs = self._func_dict["first"]
+            if isinstance(funcs, list):
+                for f in funcs:
+                    input_tensor = f(input_tensor)
+            else:
+                input_tensor = funcs(input_tensor)
+
+        for module in self._module_list:
+            input_tensor = module(input_tensor)
+            if id(module) in self._func_dict:
+                funcs = self._func_dict[id(module)]
+                if isinstance(funcs, list):
+                    for f in funcs:
+                        input_tensor = f(input_tensor)
+                else:
+                    input_tensor = funcs(input_tensor)
+
+        return input_tensor
+
+
+class LayerSpec:
+
+    def __init__(self, typename, *module_args, **module_kwargs):
+        self.typename = typename
+        self.module_args = module_args
+        self.module_kwargs = module_kwargs
+        self.children = None
+        self._param_count = 0
+
+        if not issubclass(typename, torch.nn.Module):
+            raise RuntimeError('LayerSpec only supports torch.nn.Module types.')
+
+    def __repr__(self):
+        return call_to_str(self.typename.__name__, self.module_args, self.module_kwargs)
+
+    @property
+    def param_count(self):
+        return self._param_count
+
+    def build(self):
```

Comment:
can you handle the case where a param sharded by two submodules?

Reply:
If the "build" process happens under the ZeroInit context, params will be sharded correctly.

Reply:
Sorry, I mean shared (not sharded) by two submodules.